### PR TITLE
[runtime] Handle RuntimeWrappedException and dynamic methods. Fixes #59334

### PR DESCRIPTION
**Pull request by kumpera**
Introduce MonoDynamicMethod that holds the assembly of a DM (and TB).
Use it to determine the RuntimeWrappedException policy for exceptions.

This is needed cuz the assembly of a MonoMethod is normally found using method->parent->image->assembly.

But this doesn't work with DynamicMethods that don't define an owner type as they are forced to corlib's policy.

The default policy for DM's is to not wrap those exceptions and in the case a Module is provided, the policy of that module must be respected.

This commit handles all cases but when a Module is a SRE module.

### CODE REVIEW

**Comment by kumpera** [comment]
This PR doesn't handle the SRE case, I'll do it next month when I'm back from FTO.
In case someone wants to finish it, here's the testing code I'm currently using:
```
using System;
using System.Reflection;
using System.Reflection.Emit;
using System.Runtime.CompilerServices;

class Program
{
	static Func<int> EmitDelegate (DynamicMethod dm) {
		ILGenerator il = dm.GetILGenerator ();
		var ret_val = il.DeclareLocal (typeof (int));
		var leave_label = il.DefineLabel ();

		//ret = 1;
		il.Emit (OpCodes.Ldc_I4, 1);
		il.Emit (OpCodes.Stloc, ret_val);

		// try {
		il.BeginExceptionBlock ();
		//	throw "hello";
		il.Emit (OpCodes.Ldstr, "hello");
		il.Emit (OpCodes.Throw, typeof (string));
		//	ret = 2
		il.Emit (OpCodes.Ldc_I4, 2);
		il.Emit (OpCodes.Stloc, ret_val);
		// }
		il.Emit (OpCodes.Leave, leave_label);
		//catch (string)
		il.BeginCatchBlock (typeof (string));
		il.Emit (OpCodes.Pop);
		//	ret = 3
		il.Emit (OpCodes.Ldc_I4, 3);
		il.Emit (OpCodes.Stloc, ret_val);
		//}
		il.Emit (OpCodes.Leave, leave_label);
		il.EndExceptionBlock ();

		il.MarkLabel (leave_label);
		//return ret;
		il.Emit (OpCodes.Ldloc, ret_val);
		il.Emit (OpCodes.Ret);

		var dele = (Func<int>)dm.CreateDelegate (typeof (Func<int>));
		return dele;
	}
    static int Main(string[] args)
    {
		
		AssemblyBuilder ab = AppDomain.CurrentDomain.DefineDynamicAssembly (new AssemblyName ("ehatevfheiw"), AssemblyBuilderAccess.Run);
		AssemblyBuilder ab2 = AppDomain.CurrentDomain.DefineDynamicAssembly (new AssemblyName ("ehatevfheiw"), AssemblyBuilderAccess.Run);
		CustomAttributeBuilder cab = new CustomAttributeBuilder (
				typeof (RuntimeCompatibilityAttribute).GetConstructor (new Type [0]),
				new object [0],
				new PropertyInfo[] { typeof (RuntimeCompatibilityAttribute).GetProperty ("WrapNonExceptionThrows") },
				new object[] { true });
		ab2.SetCustomAttribute (cab);
		
		DynamicMethod invoke_no_module = new DynamicMethod("throw_1", typeof (int), new Type [0]);
		DynamicMethod invoke_with_module = new DynamicMethod("throw_2", typeof (int), new Type [0], typeof (Program).Module);
		DynamicMethod invoke_with_ab = new DynamicMethod("throw_3", typeof (int), new Type [0], ab.ManifestModule);
		DynamicMethod invoke_with_ab2 = new DynamicMethod("throw_4", typeof (int), new Type [0], ab2.ManifestModule);

		try {
			int res = EmitDelegate (invoke_no_module)();
			Console.WriteLine ("GOT VAL {0}", res);
			if (res != 3)
				return 1;
		} catch (RuntimeWrappedException e) {
			Console.WriteLine ("GOT RuntimeWrappedException");
			return 2;
		}

		try {
			int res = EmitDelegate (invoke_with_module)();
			Console.WriteLine ("GOT VAL {0}", res);
			return 3;
		} catch (RuntimeWrappedException e) {
			Console.WriteLine ("GOT RuntimeWrappedException");
		}

		try {
			int res = EmitDelegate (invoke_with_ab)();
			Console.WriteLine ("GOT VAL {0}", res);
		} catch (RuntimeWrappedException e) {
			Console.WriteLine ("GOT RuntimeWrappedException");
			return 3;
		}


		try {
			int res = EmitDelegate (invoke_with_ab2)();
			Console.WriteLine ("GOT VAL {0}", res);
			return 3;
		} catch (RuntimeWrappedException e) {
			Console.WriteLine ("GOT RuntimeWrappedException");
		}
		return 0;
    }
}
```

**Comment by vargaz** [comment]
Needs a test.

**Comment by kumpera** [comment]
PR is ready for reviewing. Added test and implemented AssemblyBuilder support.

**Comment by marek-safar** [comment]
build

File: mono/metadata/sre.c [hunk 1]
```diff
@@ -4372,6 +4378,18 @@ ves_icall_AssemblyBuilder_basic_init (MonoReflectionAssemblyBuilder *assemblyb)
 }
 
 void
+ves_icall_AssemblyBuilder_UpdateNativeCustomAttributes (MonoReflectionAssemblyBuilderHandle assemblyb)
```
Comment (lambdageek) [inline mono/metadata/sre.c]: This has a final `MonoError *error` argument.

File: mono/metadata/sre.c [hunk 2]
```diff
@@ -4372,6 +4378,18 @@ ves_icall_AssemblyBuilder_basic_init (MonoReflectionAssemblyBuilder *assemblyb)
...
+{
+	MonoArrayHandle cattrs = MONO_HANDLE_NEW_GET (MonoArray, assemblyb, cattrs);
+
+	MonoReflectionAssemblyHandle assembly_handle = MONO_HANDLE_CAST (MonoReflectionAssembly, assemblyb);
+	MonoAssembly *assembly = MONO_HANDLE_GETVAL (assembly_handle, assembly);
+	g_assert (assembly);
+
+	mono_save_custom_attrs (assembly->image, assembly, MONO_HANDLE_RAW (cattrs));
```
Comment (lambdageek) [inline mono/metadata/sre.c]: Add a `/* FIXME */` comment so we eventually come back to `mono_save_custom_attrs`

**Comment by kumpera** [comment]
Linux i386 error is unrelated. FullAOT is a bug

**Comment by kumpera** [comment]
Fixed the two issues you pointed out, lets hope this fixes Linux x64

**Comment by kumpera** [comment]
I have no idea on the Windows 64 build failure.
OSX 64bits is failing on thread-suspend-selfsuspended.exe

**Pull request merged**